File: python/hashing/trappingRainWater.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        if height is None or height == []:
            return 0
        


        rv = 0
        l = 0
        r = len(height) - 1
        
        leftMax = height[l]
        rightMax = height[r]

        while l < r:
            if leftMax < rightMax:
                l += 1
                leftMax = max(leftMax, height[l])
                rv += leftMax - height[l]
            else:
                r -= 1
                rightMax = max(rightMax, height[r])
                rv += rightMax - height[r]
            

        return rv


            

print(Solution().trap([0,1,0,2,1,0,1,3,2,1,2,1]))

                    

File: java/trappingRainWater.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int maxLeft = height[0];
        int maxRight = height[n - 1];
        int l = 0, r = n - 1;
        int rv = 0;
        while (l < r)
        {
            if (height[l] < height[r])
            {
                if (height[l] > maxLeft)
                {
                    maxLeft = height[l];
                }
                else
                {
                    rv += maxLeft - height[l];
                }
                l++;
            }
            else
            {
                if (height[r] > maxRight)
                {
                    maxRight = height[r];
                }
                else
                {
                    rv += maxRight - height[r];
                }
                r--;
            }
        }
        return rv;
    }
}   